### ISSUE TITLE
paths wildcard validation seems off

### ISSUE THREAD

**Issue by bplessis-swi** [mpalmer/action-validator#27]
Hi,

Sorry if i missed something but when following the "filter pattern cheat cheet" of github action, the validator fail the file with the following message:
`Glob "**.md" in on.push.paths is invalid: Pattern syntax error near position 2: recursive wildcards must form a single path component`

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#patterns-to-match-file-paths
![image](https://user-images.githubusercontent.com/117345912/217491325-f745c8bd-02a1-4dba-96e0-5a28fb39d670.png)

Sample use:
```
on:
  push:
    branches:
      - main
    paths:
      - '**.md'
    paths-ignore:
      - '**/*.md'
```

**Comment by mpalmer** [comment]
Thanks for this solid report.  I'd say this comes down to a difference of opinion between the globbing library that action-validator uses, and what GitHub specifies.  Might not be trivial to solve, but I'll get it figured out one way or another.

**Comment by bplessis-swi** [comment]
Yeah the rust glob() function seem more "limited" in it's patterns, the equivalent of '**.md' seem to be '**/*.md' (that apparently github also understand apparently, given the 'docs/**/*.md' sample a few lines down in the same documentation.

**Comment by jhberges** [comment]
FWIW: The tool will also error on exclusion patterns.

Example:
```yaml
on:
  push:
    branches:
      - main
    paths:
      - '!generated/**'
```

Adding the extra `/*` (that is `'!generated/**/*'`) does not work around this.

I like the idea though :-)

**Comment by kevinlul** [comment]
I encountered this behaviour as well: https://github.com/DawnbrandBots/yaml-yugi/actions/runs/5626963939/job/15248759792

Claims the `directory/**` pattern is invalid, but these workflows trigger and run just fine:
- https://github.com/DawnbrandBots/yaml-yugi/blob/f55d872968ec5b400fa29a88aef2dd5865e97302/.github/workflows/validate-assignments.yaml#L5-L14
- https://github.com/DawnbrandBots/yaml-yugi/blob/f55d872968ec5b400fa29a88aef2dd5865e97302/.github/workflows/validate-data.yaml#L16

**Comment by vitalii-buchyn-exa** [comment]
Hey guys,

Experiencing a similar issue with `somedir/**` path.

Error example:
```
errors: [
        NoFilesMatchingGlob {
            code: "glob_not_matched",
            detail: Some(
                "Glob \"terraform/dev/**\" in /on/pull_request/paths does not match any files",
            ),
            path: "/on/pull_request/paths",
            title: "Glob does not match any files",
        },
    ],
```
Any workaround for that?
Thanks.

action-validator version 0.5.4
installed using asdf-vm/actions/install@v3

**Comment by DilepDev** [comment]
Experiencing a similar issue for below ci.yml

```
name: CI Workflow

on:
  push:
    paths-ignore:
      - '.gitattributes'
      - '**.MD'
      - '.devcontainer/**'
```

```
Treating ci.yml as a Workflow definition
Fatal error validating .github/workflows/ci.yml
Validation failed: ValidationState {
    action_type: Some(
        Workflow,
    ),
    file_path: Some(
        ".github/workflows/ci.yml",
    ),
    errors: [
        InvalidGlob {
            code: "invalid_glob",
            detail: Some(
                "Glob \"**.MD\" in /on/push/paths-ignore is invalid: Pattern syntax error near position 2: recursive wildcards must form a single path component",
            ),
            path: "/on/push/paths-ignore",
            title: "Glob does not match any files",
        },
        NoFilesMatchingGlob {
            code: "glob_not_matched",
            detail: Some(
                "Glob \".devcontainer/**\" in /on/push/paths-ignore does not match any files",
            ),
            path: "/on/push/paths-ignore",
            title: "Glob does not match any files",
        },
    ],
}
```


But when ```.devcontainer/**``` is replaced with ```.github/**``` ```glob_not_matched``` error related to this disappears

```
name: CI Workflow

on:
  push:
    paths-ignore:
      - '.gitattributes'
      - '**.MD'
      - '.github/**'
```

```
Treating ci.yml as a Workflow definition
Fatal error validating .github/workflows/ci.yml
Validation failed: ValidationState {
    action_type: Some(
        Workflow,
    ),
    file_path: Some(
        ".github/workflows/ci.yml",
    ),
    errors: [
        InvalidGlob {
            code: "invalid_glob",
            detail: Some(
                "Glob \"**.MD\" in /on/push/paths-ignore is invalid: Pattern syntax error near position 2: recursive wildcards must form a single path component",
            ),
            path: "/on/push/paths-ignore",
            title: "Glob does not match any files",
        },
    ],
}
```

**Comment by dabrahams** [comment]
IMO the right thing to do is disable glob validation until it can be made to match what GitHub uses.

**Comment by mpalmer** [comment]
I'm including this issue in my [action-validator code fund experiment](https://hezmatt.org/~mpalmer/code-fund.html). The code fund is designed to sustainably crowd-fund open source development in action-validator. All contributions are very much appreciated.

**Comment by ferdnyc** [comment]
> No workaround available, `action-validator` needs to be modified to use a different glob expansion mechanism in order to support these patterns.

Indeed, the globbing package currently used (the creatively named `"glob"`) is explicitly documented as offering:

```console
$ cargo info glob
glob
Support for matching file paths against Unix shell style patterns.
```

Shell patterns don't treat `**` as a "match anything" wildcard (a superset of `*`), which is what GitHub documents their `**` as meaning.

A `**` in the middle of a shell path (`.../**/`) will recursively expand to all directories at or below that location. A `**` at the _end_ of a path just becomes `*`. It NON-recursively matches everything at that level, but nothing below.

(In zsh...)
```console
$ cd .../action-validator

$ ls src
config.rs  lib.rs   schemas.rs   system    validation_error.rs
js         main.rs  schemastore  utils.rs  validation_state.rs

$ ls -d src/*
src/config.rs  src/schemas.rs    src/validation_error.rs
src/js/        src/schemastore/  src/validation_state.rs
src/lib.rs     src/system/
src/main.rs    src/utils.rs

# Identical to previous (non-recursive)
$ ls -d src/**
src/config.rs  src/schemas.rs    src/validation_error.rs
src/js/        src/schemastore/  src/validation_state.rs
src/lib.rs     src/system/
src/main.rs    src/utils.rs

# Actually recursive
$ ls -d src/**/*
src/config.rs     src/schemas.rs         src/system/mod.rs
src/js/           src/schemastore/       src/system/process.rs
src/js/system.js  src/system/            src/utils.rs
src/lib.rs        src/system/console.rs  src/validation_error.rs
src/main.rs       src/system/fs.rs       src/validation_state.rs
```

(In bash, which interestingly doesn't allow `**/*` to match directories directly alongside the `**`; IOW, `/a/b/**/*` doesn't decay to include `/a/b/*` like in zsh, the `**` has to match at least one directory level...)
```console
$ cd .../action-validator

$ ls src
config.rs  lib.rs   schemas.rs   system    validation_error.rs
js         main.rs  schemastore  utils.rs  validation_state.rs

$ ls -d src/*
src/config.rs  src/schemas.rs   src/validation_error.rs
src/js         src/schemastore  src/validation_state.rs
src/lib.rs     src/system
src/main.rs    src/utils.rs

# Identical to previous (non-recursive)
$ ls -d src/**
src/config.rs  src/schemas.rs   src/validation_error.rs
src/js         src/schemastore  src/validation_state.rs
src/lib.rs     src/system
src/main.rs    src/utils.rs

# Actually recursive, but consumes a path level
$ ls -d src/**/*
src/js/system.js       src/system/fs.rs   src/system/process.rs
src/system/console.rs  src/system/mod.rs
```

I think the Rust glob package is still buggy, because for example glob validation will claim that a glob like `src/system/**` matches no files in this repo, when clearly even the shells think it does:

```console
$ ls src/system/**
src/system/console.rs  src/system/mod.rs
src/system/fs.rs       src/system/process.rs
```

...But that's a moot point, because the package's goal is to replicate shell globbing. Even if it did that perfectly, GitHub's globbing is different and more flexible.

**Comment by mpalmer** [comment]
Yes, it's generally agreed that it'll be necessary to switch to a different globbing package (or build one explicitly to mimic GitHub's behaviours) in order to solve this issue properly.  If you've got the available time and energy, identifying which globbing implementation most closely matches what GitHub does (or being able to say with some degree of authority "they're all totally wrong") would be extremely helpful.

**Comment by ferdnyc** [comment]
I've started looking at them. [wax](https://docs.rs/wax/0.6.0/wax/) seems the most mature and promising so far, but its docs don't seem to explicitly say how _exactly_ it handles `**` on the _end_ of a pattern. Guess there's nothing for it but to try!

**Comment by ferdnyc** [comment]
I take that back. [One of its tests](https://docs.rs/wax/0.6.0/src/wax/lib.rs.html#1635-1647) makes it clear that a wax glob like `a/b/**` is at least expected to match `a/b/file.ext`, so that's already a step in the right direction. (The question is, does it also match `a/b/c/file.ext`?)

**Comment by anttiharju** [comment]
> (or being able to say with some degree of authority "they're all totally wrong") would be extremely helpful.

Went ahead and tested top-5 libraries when searching 'glob' from crates.io with https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#patterns-to-match-file-paths. Results:


https://crates.io/crates/glob is not a fit, because `'*'` will match `docs/file.txt`.  
https://crates.io/crates/wax is not a fit, because `*.jsx?` will not match `page.js` or `page.jsx`.  
https://crates.io/crates/globset is not a fit, because `'*'` will match `'docs/file.txt'`  
https://crates.io/crates/fast-glob is not a fit, because `'*.jsx?'` will not match `page.js`.  
https://crates.io/crates/glob-match is not a fit, because `'*.jsx?'` will not match `page.js`.

=> "they're all totally wrong"

**Comment by anttiharju** [comment]
Began work on a library [`pathglob`](https://github.com/anttiharju/pathglob/blob/fb984a9d4667455e9ca9e2250afe45ee301a0431/docs/README.md). It's nowhere near ready but just sharing progress.

I'll try to implement the necessary changes in action-validator to also inform library design.

**Comment by anttiharju** [comment]
Seeing some initial success at https://github.com/anttiharju/action-validator/pull/3

still needs a lot of work

**Comment by anttiharju** [comment]
There's now a PR open to fix this:
* https://github.com/mpalmer/action-validator/pull/113

**Comment by mpalmer** [comment]
Closed via #113.
